[docs] Add SSR section (#880)
lit/lit.dev: packages/lit-dev-content/site/docs/ssr/server-api.md

---
title: Lit SSR server API
eleventyNavigation:
  key: Server API
  parent: Server-side rendering
  order: 2
---

> Lit SSR provides a couple different ways of rendering custom elements server-side: rendering in the [global scope](#global-scope) or via [VM modules](#vm-module).

Author: I'm unsure about focusing purely on how the DOM shim is loaded as being the sole differentiator between the two ways since there may be other reasons to prefer isolation.

My original take was that we have two ways of rendering, global and isolated in VM modules where how the DOM shim is loaded happens to be one difference, albeit a significant one, between the two.

All of our existing SSR doc and code refer to adding the DOM APIs as "DOM Shim". @justinfagnani suggested a section called "DOM Compatibility (or DOM Emulation)". In this draft I've been using them interchangeably though maybe we should just stick to DOM Shim consistently instead?

With regards to VM modules, perhaps we can say something along the lines of
`The 'renderModule()` function utilizes Node's [`vm.Module` class](https://nodejs.org/api/vm.html#class-vmmodule) which enables running code within V8 Virtual Machine contexts.

> Lit SSR provides a couple different ways of rendering custom elements server-side: rendering in the [global scope](#global-scope) or via [VM modules](#vm-module).

Reviewer: OK. Perhaps I missed something in reading the section, I just didn't see any other differences mentioned. You could soften my earlier suggestion a bit like this:

```suggestion
Lit SSR provides two ways of rendering custom elements server-side: rendering in the [global scope](#global-scope) or using[VM modules](#vm-module). The two methods differ primarily in how the DOM shim is loaded.
```

I don't think the section title is a problem, but if you use both DOM emulation and DOM shim in this section, I'd just try to make it really clear that you were talking about the same thing.

In order to render custom elements in Node, Lit SSR requires some minimal [DOM emulation]('/docs/ssr/dom-emulation') necessary for `lit-html` and `reactive-element` to work, as well as a working `CustomElementRegistry`.

When rendering in the global scope, this DOM shim is installed in the global space. This may cause unintended behaviors for other libraries which, for instance, might try to detect the running environment by checking for the presence of `window` in the global space. All render requests also share the same global custom element registry, and any data that might be stored globally.

> Rendering with VM modules allow each render request to have its own context with a separate global from the main Node process such that the DOM emulation is only installed within that scope and any custom elements being registered or global data is contained within that context used only for that particular render request. It does require utilizing an experimental node feature with a particular way of using it. See below for details.

Author: I'd say the usage is slightly less idiomatic since instead of calling a function directly, you have to set up a separate module file, with a function within it, and specify the function name to call within that. Perhaps it's best to leave that portion to the table though.

> Rendering with VM modules allow each render request to have its own context with a separate global from the main Node process such that the DOM emulation is only installed within that scope and any custom elements being registered or global data is contained within that context used only for that particular render request. It does require utilizing an experimental node feature with a particular way of using it. See below for details.

Reviewer: Got it. Yeah, I'd say either leave it to the table or be a little more concrete, like, "Rendering with VM modules requires a little extra boilerplate code."

---
title: Lit SSR server API
eleventyNavigation:

>   key: Server API

Reviewer 2: Change to "Server Usage"

  parent: Server rendering
  order: 2
---

{% labs-disclaimer %}

In order to render custom elements in Node, they must first be defined and registered with the global `customElements` API, which is a browser-only feature. As such, Lit SSR includes a DOM shim that provides the minimal DOM APIs necessary to render Lit on the server. (For a list of emulated APIs, see [DOM emulation](/docs/ssr/dom-emulation).)

> Lit SSR provides two different ways of rendering custom elements server-side: rendering in the [global scope](#global-scope) or via [VM modules](#vm-module), which utilizes Node's [`vm.Module`](https://nodejs.org/api/vm.html#class-vmmodule) which enables running code within V8 Virtual Machine contexts. The two methods differ primarily in how the DOM shim is loaded.

Reviewer 3: ```suggestion
Lit SSR provides two different ways of rendering custom elements server-side: rendering in the [global scope](#global-scope) or via [VM modules](#vm-module). VM modules utilizes Node's [`vm.Module`](https://nodejs.org/api/vm.html#class-vmmodule) API, which enables running code within V8 Virtual Machine contexts. The two methods differ primarily in how the DOM shim is loaded.
```

When rendering in the global scope, shimmed DOM globals (like `window`, `HTMLElement`, `customElements`, etc.) are added directly to the Node.js global scope. This may cause unintended behaviors for other libraries. For instance, some libraries try to detect the running environment by checking for the presence of `window` in the global scope. In addition, all render requests also share the same global custom element registry, and any data that might be stored globally.

Rendering with VM modules allows each render request to have its own context with a separate global from the main Node process. DOM emulation is only installed within that context, any custom elements being registered are registered in that context, and any global data is contained within that context. VM modules are an experimental Node feature.

| Global | VM Module |
|-|-|

> | Pros:<ul><li>Easy to use–can import component modules directly and call `render()` with templates.</li></ul>Cons:<ul><li>Introduces DOM shim to global scope, potentially causing issues with other libraries.</li><li>Custom elements are registered in a shared registry across different render requests.</li></ul> | Pros:<ul><li>Does not introduce DOM shim to the global scope.</li><li>Isolates contexts across different render requests.</li></ul>Cons:<ul><li>Less intuitive usage–need to write and specify a module file with a function to call.</li><li>Slower due the module graph needing to be re-evaluated per request.</li></ul> |

Reviewer 3: ```suggestion
| Pros:<ul><li>Easy to use. Can import component modules directly and call `render()` with templates.</li></ul>Cons:<ul><li>Introduces DOM shim to global scope, potentially causing issues with other libraries.</li><li>Custom elements are registered in a shared registry across different render requests.</li></ul> | Pros:<ul><li>Does not introduce DOM shim to the global scope.</li><li>Isolates contexts across different render requests.</li></ul>Cons:<ul><li>Less intuitive usage. Need to write and specify a module file with a function to call.</li><li>Slower due the module graph needing to be re-evaluated per request.</li></ul> |
```

I find the dashes a little weird to read here -- it makes me think I'm reading a hyphenated word "use-can".

---
title: Lit SSR server API
eleventyNavigation:
  key: 🧪 Server API
  parent: Server-side rendering
  order: 2
---

<details class="pre-release">
  <summary> 🧪 Pre-release software: not for production.</summary>
  

>   Lit labs packages are published to get feedback from the wider community. This code shouldn't be used in production, but we encourage you to try it out and [file issues](https://github.com/lit/lit/issues/new/choose) for any bugs you find. For general feedback, please use the GitHub [discussion](https://github.com/lit/lit/discussions).

Reviewer 2: We should make a shortcode/ helper so this isn't repeated and can be updated centrally.

Reviewer 4: Was gonna say the same thing; we have a tracking issue in https://github.com/lit/lit.dev/issues/838) I also think we should use a more more general statement with a link to a more detailed description on the Labs page (rather than repeating the caveats on each page that references them), something like:

> This package is part of the Lit Labs family of experimental packages. See the [Lit Labs](#) page for guidance on using Labs software in production."

And I think we need to workshop the contents of the detailed description. Specifically, I'm not sure about saying "This code _shouldn't_ be used in production", because I think we do want to encourage not just tire kicking but actual real-world usage; the caveat is that it needs to be done with full understanding that breaking changes are more likely, and comes with the possibility of being end-of-lifed if it doesn't end up being generally useful enough.

For more information about the Lit labs process, see [Lib Labs](/docs/libraries/labs/)

</details>

> Lit SSR provides two different ways of rendering custom elements server-side: rendering in the [global scope](#global-scope) or via [VM modules](#vm-module), which utilizes Node's [`vm.Module`](https://nodejs.org/api/vm.html#class-vmmodule) which enables running code within V8 Virtual Machine contexts. The two methods differ primarily in how the DOM shim is loaded.

Reviewer 2: What about workers?

Author: We use workers for the eleventy plugin but don't provide anything for users to use workers from labs/ssr package yet. They'd have to manually run scripts in worker threads that utilize the render with global dom shim.

> In order to render custom elements in Node, Lit SSR includes a DOM shim that provides the minal DOM APIs necessary to render Lit on the server. (For a list of emulated APIs, see [DOM emulation](/docs/ssr/dom-emulation).)

Reviewer 4: Per your other comment above about the DOM shim not being the _only_ difference, I think it could be addressed by saying

> In order to render custom elements in Node, they must first be defined and registered with the global `customElements` API, which is a browser-only feature. As such, Lit SSR includes a DOM shim that provides minimal DOM APIs necessary to register and render elements in a nodejs environment. (For a list of emulated APIs, see [DOM emulation](/docs/ssr/dom-emulation).)
